README
Finance-InteractiveBrokers-SWIG

This module provides Perl connectivity to the InteractiveBrokers market data
and program trading service, using the IB-provided C++ code.  It is primarily
intended to be used with L<POE::Component::Client::InteractiveBrokers>, which
provides a better API, but may be used standalone if desired, by referring
to the IB documentation itself (under L</"SEE ALSO">).

It is a very complex module with an involved build process, and thus you
should read the documentation thoroughly.


PREREQUISITES

You must have the following to build and use this module:

* Finance::InteractiveBrokers::API

    Provides a programmatic means of looking up methods and events in the
    IB API.

* A working build environment

    Capable of compiling C and C++ files, and running 'make'.

* SWIG >= 1.3.28

    The "Simplified Wrapper and Interface Generator", capable of building SWIG
    interfaces.  This module has been tested with versions from 1.3.28-2.0.1.

Optional, but highly recommended:

* Alien::InteractiveBrokers

    Installs (downloading if necessary) the InteractiveBrokers API files,
    and provides Perly mechanisms for locating them.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install


HOW IT WORKS

The IB API is available as either a set of C++ or Java source code files.
This module builds a library from this, and then runs SWIG (the Simplified
Interface Wrapper and Generator) against it to provide Perl connectivity.

The API consists of several methods, callable from this module, as well
as several events, containing the asynchronous responses from IB to
methods you have called.

In order to catch the events, you must subclass
L<Finance::InteractiveBrokers::SWIG::EventHandler>, and override all of
the events therein with your own code to handle their responses (e.g.
save them to a database, or do whatever).

You then pass your $handler into ->new(), and you have complete access
to the IB API, delegated through the C++ library.


NOTES ON BUILDING ON OS-X

When building on OS-X, you may see warnings similar to the following, when
linking the final dynamic library:

ld: warning: directory not found for option '-L/usr/local/lib'
ld: warning: directory not found for option '-L/usr/local/lib'
ld: warning: directory not found for option '-L/usr/local/lib'
ld: warning: directory not found for option '-L/usr/local/lib'

This is avoidable, but it would reduce the flexibility and require a much
more convoluted build process.

An explanation of why this happens: When distributing modules, especially
binary modules as complex as Finance::InteractiveBrokers::SWIG for myriad
end-user platforms, the only way to get an idea of the local system's
configuration is to query Perl itself, via various methods.  The build
process expects that, since perl is built and (hopefully!) working, that
its configuration will work.  This is not the case if the people who
packaged your copy of 'perl' do not use a common system configuration such as
would be found on an OS-X end-user's machine.  In particular, it looks like
the packagers had some libraries installed in '/usr/local/lib', which is
not necessarily a normal directory on OS-X.  Blame them, not me.

The solution is to ignore these warnings; the final build works fine, if
everything else has completed successfully, and a 'make test' reports a
PASS.


SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Finance::InteractiveBrokers::SWIG

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Finance-InteractiveBrokers-SWIG

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Finance-InteractiveBrokers-SWIG

    CPAN Ratings
        http://cpanratings.perl.org/d/Finance-InteractiveBrokers-SWIG

    Search CPAN
        http://search.cpan.org/dist/Finance-InteractiveBrokers-SWIG/


LICENSE AND COPYRIGHT

Copyright (c) 2010-2012 Jason McManus

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

The authors are not associated with InteractiveBrokers, and as such, take
no responsibility or provide no warranty for your use of this module or the
InteractiveBrokers service.  You do so at your own responsibility.  No
warranty for any purpose is either expressed or implied by your use of this
module suite.

The data from InteractiveBrokers are under an entirely separate license that
varies according to exchange rules, etc.  It is your responsibility to
follow the InteractiveBrokers and exchange license agreements with the data.

# END


Changes
Revision history for Finance-InteractiveBrokers-SWIG
------------------------------------------------------------
Legend:
BUG:    An actual found or reported bug fix.
FIX:    Change toward correct behavior, but unreported.
NEW:    New, previously non-existent feature.
CHG:    Change in behavior.
API:    Change in user-facing API (hopefully rare).
DOC:    Changes to documentation.
TST:    New tests
------------------------------------------------------------

0.12     2014-06-01
         BUG: RT#96110 - RT#88097 was incomplete - thanks!
         TST: Tests for above now actually work..
         BUG: RT#94918 - Fixed examples/api-demo.pl - thanks!
         DOC: RT#94897 - IB changed their API doc URL.

0.11     2014-04-20
         BUG: RT#94880 - gcc >4.7 stopped including <unistd.h> by default
         FIX: F::IB::API version check from 0.10 was broken
         FIX: Regression tests for 0.10 were not added to MANIFEST

0.10     2013-08-31
         NEW: RT#88097 - Add setSelectTimeout() method
         TST: Added a couple of regression tests

0.09     2013-08-26
         BUG: RT#88097 - Spurious error message (IBAPI bug) - thanks!
         FIX: Refined OS-X broken compiler check - LLVM 2335 rejected

0.08     2012-08-14
         BUG: RT#79014 - SWIG min version needed to be 1.3.33 - thanks!
         TST: Added actual string regression test for the above

0.07     2012-08-04
         Bump to production release

0.06_02  2012-05-02
         BUG: RT#76932 - OS-X segfault - ExtUtils::Embed is made of lies.

0.06_01  2012-04-29
         NEW: Added support for IB API v9.65-9.67
         NEW: 9.66 supports reqMarketDataType method and marketDataType event
         NEW: 9.67 supports commissionReport event
         API: version() now returns a double (shouldn't affect Perl)

0.06     2012-04-26
         BUG: RT#76806 - Invalid event received - thanks, rod!
         BUG: RT#76807 - OrderId needed to be typedef'ed - thanks, rod!

0.05     2012-03-31
         No reported bugs on 0.04_01; bumping to an official release.

0.04_01  2011-10-13
         BUG: F::IB::SWIG swig search on win32 incorrect (thanks, leosusanto!)

0.04     2011-10-09
         No reported bugs in 0.03_04; bumping to an official release.

0.03_04  2011-09-10
         FIX: Get rid of -fstack-protector in Makefile.swig

0.03_03  2011-08-27
         FIX: Fixed Config.pm handling
         FIX: Properly imported File::Spec::Functions qw( catdir catfile )
         FIX: Proper parameters in examples/MyEventHandler.pm
         FIX: %INC handling more portable
         FIX: Add typedefs for TickerId and IBString
         FIX: Bad $LIBPATH handling in Makefile.PL
         FIX: Better standardized Makefile.swig variables
         FIX: Lowered required swig version to 1.3.28 after testing
         NEW: More thorough unit tests
         NEW: More comprehensive demo in examples/api-demo.pl
         NEW: Added hooks for Alien::SWIG CPAN d/l during build process
         BUG: F::IB::SWIG::_resolve_host misused Net::hostent (thanks, Uwe!)
         BUG: F::IB::SWIG::Makefile.swig - OS-X build borked (thanks, Chris!)
         DOC: Fixed bad function name in SWIG.pm synopsis
         DOC: Notes on OS-X building in README
         FIX: Added Test:More => 0.88 requirement, for done_testing()

0.03     2011-01-27
         FIX: Makefile.PL, META.yml needed configure_requires => F::IB::API
         NEW: examples/show-*.pl informational utilities for cmdline
         DOC: Minor doc cleanup and additions

0.02     2011-01-25
         First production CPAN release
         Provides interface for InteractiveBrokers API version 9.64
         DOC: Minor documentation tweaks
         FIX: Versioning scheme will change from here on.

0.01_02  2011-01-25     (developer release)
         DOC: Fixing documentation bugs seen on CPAN from 9.64_01

0.01_01  2011-01-24     (developer release)
         Main suite should be complete
         Provides interface for InteractiveBrokers API version 9.64
         Prepped for releasing to CPAN for cpantesters feedback

# END


Makefile.PL
#
#   Finance::InteractiveBrokers::SWIG - Perl makefile maker
#
#   Copyright (c) 2010-2014 Jason McManus
#

use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;
use File::Spec::Functions qw( catdir catfile );
use Config;
use Cwd;
use Carp qw( carp croak confess );
$|=1;

use vars qw( $VERSION $TRUE $FALSE );
BEGIN {
    $VERSION = '0.12';
}

*TRUE    = \1;
*FALSE   = \0;

my $SWIG_filename = 'swig';
$SWIG_filename .= '.exe' if( $^O eq 'MSWin32' );

my $ALIEN_SWIG = 'Alien::SWIG';
my $min_SWIG_version = '1.3.33';
my( $SWIG_PATH, $SWIG_INCLUDES ) = ( '', '' );

my @broken_compilers = (
	'LLVM build 2335',
);

# Check for F::IB::API >0.04
my $fibapi_outdated = $TRUE;
eval {
    require Finance::InteractiveBrokers::API;
    $fibapi_outdated = ($Finance::InteractiveBrokers::API::VERSION < 0.04);
};
croak( '*' x 76 . "\n" .
       '  Cannot find Finance::InteractiveBrokers::API >= 0.04, required for this' .
       " build.\n  Please install it from CPAN before continuing.\n" .
       '*' x 76 . "\n " )
    if( $@ or $fibapi_outdated );

##########################################################################
### Collect some configuration information from the user
##########################################################################

print <<"EOM";

==========================================================================
      Welcome to the Perl Finance::InteractiveBrokers::SWIG package
==========================================================================

We need to gather some information before continuing the build process.

This module is very complex, and thus you should consult the documentation
before using it.

Press ENTER at any prompt to accept the [default] value.
EOM


###########################
### Find the swig binary
###########################

print <<"EOM";

---------------------------------------
               SWIG:
---------------------------------------

This build requires SWIG (the Simplified Wrapper and Interface Generator)
for building the connection to the IB C++ libraries.

If it is not installed, you can get it at: http://www.swig.org/

EOM

print " * Looking for 'swig' in your PATH... ";
my $ver_ok;
my $swig_bin = find_bin( $SWIG_filename );
if( $swig_bin ) {
    print $swig_bin, "\n\n";
    $ver_ok = check_swig_version( $swig_bin );
} else {
    print "Nope.\n\n";
}

unless( $swig_bin and $ver_ok )
{
    print " * Trying $ALIEN_SWIG... ";
    eval "require $ALIEN_SWIG;";
    if( $@ ) {
        print "Nope.\n\n";
    } else {
        my $aswig = $ALIEN_SWIG->new();
        print "Found it!  $ALIEN_SWIG ", $ALIEN_SWIG->VERSION, "\n\n";
        $swig_bin = $aswig->executable();
        $ver_ok = check_swig_version( $swig_bin );
        $SWIG_INCLUDES = $aswig->includes()
            if( $ver_ok );
    }
}

# Find and verify if location of SWIG binary
$swig_bin  = '' unless( defined( $swig_bin ) );
$SWIG_PATH = prompt( "Enter/change the full path to your 'swig' program:\n>",
                        $ver_ok ? $swig_bin : undef );

print "\n";
unless( ( ( $SWIG_PATH ) and ( $SWIG_PATH eq $swig_bin ) ) or
        ( $ver_ok = check_swig_version( $SWIG_PATH ) ) )
{
    print "No good I'm afraid.  ";
    print "We can try to get $ALIEN_SWIG from CPAN.\n\n";
    my $get_it = prompt( "Download $ALIEN_SWIG from CPAN?", 'yes' );
    ( $SWIG_PATH, $SWIG_INCLUDES ) = get_aswig_from_cpan()
        if( $get_it =~ /y/ );

    die "A working SWIG >= $min_SWIG_version is needed.  Cannot continue.\n "
        unless( $SWIG_PATH );

    print "\n*********************************************************************\n";
    print "  Welcome back to the Finance::InteraactiveBrokers::SWIG installer.\n";
    print "SWIG ", _get_swig_ver( $SWIG_PATH ),
          " is now installed and working properly.\n";
    print "*********************************************************************\n";
}

###########################
### Find the API files
###########################

print <<"EOM";

---------------------------------------
  InteractiveBrokers API Source Code:
---------------------------------------

To build this module, you must have the InteractiveBrokers API source code
somewhere on your system.

We're going to try to locate it.

EOM

my( $LIBPATH, $IB_BUILD_PATH );

print " * Checking the IB_LIBRARY_PATH and IB_BUILD_PATH envvars... ";

if( ( $LIBPATH = found_api( $ENV{IB_BUILD_PATH} ) ) or
    ( $LIBPATH = found_api( $ENV{IB_LIBRARY_PATH} ) ) )
{
    print "Found it!\n";
}
else
{
    # Next, try to find it via ibapi-config
    print "Nope.\n\n * Trying 'ibapi-config --path'... ";
    undef $LIBPATH;
    if( my $ibapiconfig = find_bin( 'ibapi-config' ) )
    {
        $LIBPATH = `$ibapiconfig --path`;
        chomp $LIBPATH;
    }
    if( found_api( $LIBPATH ) )
    {
        print "Found it!\n";
    }
    else
    {
        # Next, try to find it via Alien::InteractiveBrokers
        print "Nope.\n\n * Checking if we can use Alien::InteractiveBrokers... ";
        undef $LIBPATH;
        eval {
            require Alien::InteractiveBrokers;
            $LIBPATH = Alien::InteractiveBrokers::path();
        };
        if( found_api( $LIBPATH ) )
        {
            print "Yep!\n";
        }
        else
        {
            print "Nope.\n";
            undef $LIBPATH;
        }
    }
}

if( $LIBPATH )
{
    my $api_ver_num = get_api_version( $LIBPATH );
    print <<"EOM";

We have found the IB API source (version $api_ver_num) in:
  $LIBPATH

You may change locations now if you need to (otherwise just press ENTER).

EOM
} else { print "\n" }

my $NEWLIBPATH = prompt( "Enter or confirm the path to the IB source code:\n", 
                         $LIBPATH );

# http://www.cpantesters.org/cpan/report/9130b39c-2c49-11e0-b43e-04429eb006f9
# FIXME $LIBPATH can be undef here.
if( ( $NEWLIBPATH ne $LIBPATH ) and
    ( ! found_api( $NEWLIBPATH ) ) )
{
    print <<"EOM";

Could not find the IB source code in $NEWLIBPATH !
This package requires the header files provided by InteractiveBrokers.
Please read this module's documentation for more information.

EOM
    die "Cannot continue without IB source code..";
}
else
{
    # Check if we can use that API version
    can_use_api_version( get_api_version( $NEWLIBPATH ) );

    print <<"EOM";

IB_BUILD_PATH has been set to '$NEWLIBPATH'.

EOM
    # Now set the env to whatever they entered
    $IB_BUILD_PATH = $NEWLIBPATH
}


##########################################################################
### Build the Makefile
##########################################################################

#$Verbose = 1;
WriteMakefile(
    NAME                => 'Finance::InteractiveBrokers::SWIG',
    AUTHOR              => q{Jason McManus <infidel AT cpan.org>},
    VERSION_FROM        => 'lib/Finance/InteractiveBrokers/SWIG.pm',
    ABSTRACT_FROM       => 'lib/Finance/InteractiveBrokers/SWIG.pm',
    ($ExtUtils::MakeMaker::VERSION >= 6.3002
      ? ('LICENSE'=> 'perl')
      : ()),
    PL_FILES            => {},
    PREREQ_PM => {
		# For done_testing()
        'Test::More'                        => '0.88',
        'Finance::InteractiveBrokers::API'  => '0.04',
    },
    BUILD_REQUIRES => {
        'Finance::InteractiveBrokers::API'  => '0.02_03',
        # Catch CPANTesters and install the IB libs
        # XXX: Nah, this won't work, already too late for 'make dist'
        #$ENV{AUTOMATED_TESTING} ? (
        #   'Alien::InteractiveBrokers'     => 0,
        #) : (),
        'Alien::InteractiveBrokers'         => 0,
    },
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker'               => 0,
        'Finance::InteractiveBrokers::API'  => 0,
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES =>
      'Finance-InteractiveBrokers-SWIG-* *.old IBAPI.pm IBAPI.so IBAPI.bundle IBAPI_wrap.cxx'
    },
    test => {
		TESTS => 't/*.t t/regression/*.t'
	},
);

print <<"EOM";

You may now continue the build by typing:

    make
    make test
    make install

Please consult the documentation via:

    perldoc Finance::InteractiveBrokers::SWIG

EOM

##########################################################################
### Utility subs
##########################################################################

# Retrieve a var from the ginormous %Config hash
sub get_config_var
{
    my $var = shift;

    return exists( $Config{$var} )
             ? $Config{$var}
             : undef;
}

# Find the specified $fname in all dirs in $ENV{PATH}
sub find_bin
{
    my( $fname ) = @_;

    my $path_sep = get_config_var( 'path_sep' ) || ':';

    my @bindirs = split( $path_sep, $ENV{PATH} );

    my $binpath = '';
    for my $dir ( @bindirs )
    {
        my $testpath = catfile( $dir, $fname );
        $binpath = $testpath, last
            if( -f $testpath );
    }

    return( $binpath );
}

# Get the SWIG version number from the executable
sub check_swig_version
{
    my( $swig ) = @_;

    return( $FALSE )
        unless( -x $swig );

    print "Checking SWIG version... ";

    my $ver = _get_swig_ver( $swig );

    unless( defined( $ver ) ) {
        print "I don't think that '$swig' is actually SWIG.\n\n";
		return( $FALSE );
    }

    print $ver, "\n\n";

    my $ok = $ver ge $min_SWIG_version;
    printf "SWIG $ver is %srecent enough.%s\n\n",
           $ok ? '' : 'NOT ',
           $ok ? '  Great!' : '  You need a newer version.';

    return( $ver ge $min_SWIG_version ? $TRUE : $FALSE );
}

# Low-level, just execute and grab the ver.
sub _get_swig_ver
{
    my( $path ) = @_;

    my $output = eval { qx( $path -version ) };
    return if( $@ or ! defined( $output ) );
    my( $ver ) = ( $output =~ m/SWIG Version (.*)/m );
    chomp $ver if( defined( $ver ) );

    return( $ver );
}

# Check a few needed files in the IBJts path to see if it's sane
sub found_api
{
    my( $path ) = @_;
    return unless( $path );

    my $result = undef;
    if( -e $path )
    {
        # Check some files to make sure we're kinda complete
        for my $file (
            'cpp',
            catdir(  'cpp', 'Shared' ),
            catfile( 'cpp', 'Shared', 'EWrapper.h' ),
            catdir(  'cpp', 'PosixSocketClient' ),
            catfile( 'cpp', 'PosixSocketClient', 'EClientSocketBase.cpp' ),
            catfile( 'cpp', 'PosixSocketClient', 'EPosixClientSocket.cpp' ) )
        {
            return unless( -e catfile( $path, $file ) );
        }
        $result = $path;
    }

    return $result;
}

# Get the IB API version number from the source code distribution
sub get_api_version
{
    my( $path ) = @_;
    return unless( $path );

    my $verfile = catfile( $path, 'API_VersionNum.txt' );
    open my $fd, '<', $verfile or
        return;
    my $contents = do { local $/; <$fd> };
    close( $fd );

    my( $vernum ) = ( $contents =~ m/API_Version=([\d.]*)/mi );

    return( $vernum );
}

# Check with F::IB::API to see if we can work with $VERSION
sub can_use_api_version
{
    my( $api_version ) = @_;

    require Finance::InteractiveBrokers::API;
    my @ok_vers = Finance::InteractiveBrokers::API->api_versions();

    die "\nSorry, this module cannot use found API version $api_version\n" .
        "Usable vers are: @ok_vers\n "
        unless( grep { $api_version } @ok_vers );

    print "\nGood, API version $api_version is usable with this module.\n";

    return;
}

# Get and install Alien::SWIG from CPAN, returning the executable and includes
sub get_aswig_from_cpan
{
	my $origpath = getcwd();
    eval "use CPAN;";
    carp "Cannot 'use CPAN': $@" and return
        if( $@ );

    CPAN->import;
    CPAN::HandleConfig->load();
    CPAN::Shell::setup_output();
    CPAN::Index->reload;

    my $mod = CPAN::Shell->expand( 'Module', $ALIEN_SWIG );
    carp "Couldn't find $ALIEN_SWIG on CPAN." and return
        unless( defined( $mod ) );

    if( ! $mod->uptodate )
    {
        $mod->install();
        carp "Couldn't install/update $ALIEN_SWIG from CPAN." and return
            unless( $mod->uptodate );
    }

	eval "require $ALIEN_SWIG;";
    carp "Couldn't require $ALIEN_SWIG after CPAN install." and return
		if( $@ );

    my $aswig = $ALIEN_SWIG->new();

	# I think something funky goes on with paths with the above.
	chdir( $origpath );

    return( $aswig->executable(), scalar $aswig->includes() );
}

sub get_cxx_fix
{
	my $quiet = shift;
	return unless $^O eq 'darwin';
	my $cxx_verstring = `g++ --version`;
	my( $cxx_version ) = $cxx_verstring =~ m#^(\S+)#;

	return unless( grep { /\Q$cxx_version\E/ } @broken_compilers );
	unless( $quiet )
	{
		print <<"EOWHINE";

*************************************************************************

	!!! BROKEN COMPILER !!!

	Your C++ compiler ($cxx_version) is BROKEN.

	The g++ on Darwin with LLVM build 2335 produces invalid output.

	You should upgrade to the latest OS-X developer tools.

EOWHINE
	}

	my $gpp = find_bin( 'g++-4.2' );

	if( -x $gpp ) {
		print <<"EOWHINE";

	Found a possible replacement: $gpp

	Will try to continue using that.

*************************************************************************
EOWHINE
		return "CC=$gpp";
	} else {
		print <<"EOWHINE";
	No suitable replacement was found.  Aborting.

*************************************************************************
EOWHINE
		die "BROKEN COMPILER; cannot continue.";
	}

	return;
}


##########################################################################
### Tweak the Makefile output
##########################################################################

package MY;

use File::Spec::Functions qw( catdir catfile );

sub top_targets {
    my $postamble;

    # I don't think we can get around much of this with File::Spec->catfile
    my $S        = $^O eq 'MSWin32' ? '\\' : '/';
    my $base_dir = catdir( 'Finance', 'InteractiveBrokers', 'SWIG');
    my $lib_dir  = "\$(INST_LIB)${S}$base_dir";
    my $auto_dir = "\$(INST_ARCHLIB)${S}" . catdir( 'auto', $base_dir, 'IBAPI');
    my $perm_dir = $ExtUtils::MakeMaker::VERSION >= 6.52 ? '$(PERM_DIR)' :'755';
    my $api_ver  = main::get_api_version( $IB_BUILD_PATH );
	my $CXX_FIX  = main::get_cxx_fix() || '';
	my $api_int  = $api_ver;
	$api_int =~ s/\.//;
    my $build_time = time();

    # Build the Makefile postamble
    $postamble = <<"POSTAMBLE";
all :: pure_all manifypods
\t\$(NOECHO) \$(NOOP)

pure_all :: config pm_to_blib swig
\t\$(NOECHO) \$(NOOP)

subdirs :: \$(MYEXTLIB)
\t\$(NOECHO) \$(NOOP)

config :: \$(FIRST_MAKEFILE) blibdirs
\t\$(NOECHO) \$(NOOP)

help :
\tperldoc ExtUtils::MakeMaker

swig :: IBAPI.\$(DLEXT)

IBAPI.\$(DLEXT) :: IBAPI.i IBAPI.cxx IBAPI.h
\t$CXX_FIX\$(MAKE) -ef Makefile.swig "SWIG=$SWIG_PATH" "SWIG_INCL=$SWIG_INCLUDES" "IB_BUILD_PATH=$IB_BUILD_PATH" "IB_API_VERSION=$api_ver" "BUILD_TIME=$build_time" "IB_PI_INTVER=$api_int"
\t\$(MKPATH) "$lib_dir" "$auto_dir"
\t\$(CHMOD) $perm_dir "$auto_dir"
\t\$(CP) IBAPI.pm "$lib_dir"${S}
\t\$(CP) IBAPI.\$(DLEXT) "$auto_dir"${S}

clean ::
\t\$(MAKE) -ef Makefile.swig clean

POSTAMBLE

    return $postamble;
}

# END
